maven/bnd-export-maven-plugin/src/main/java/aQute/bnd/maven/export/plugin/ExportMojo.java

```diff
@@ -4,82 +4,41 @@
 
 import java.io.File;
 import java.io.OutputStream;
+import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.Map.Entry;
-import java.util.Properties;
 import java.util.Set;
 
 import org.apache.maven.artifact.DefaultArtifact;
 import org.apache.maven.artifact.handler.DefaultArtifactHandler;
-import org.apache.maven.execution.MavenSession;
-import org.apache.maven.plugin.AbstractMojo;
-import org.apache.maven.plugin.MojoExecutionException;
-import org.apache.maven.plugin.MojoFailureException;
-import org.apache.maven.plugins.annotations.Component;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.Parameter;
 import org.apache.maven.plugins.annotations.ResolutionScope;
-import org.apache.maven.project.MavenProject;
-import org.apache.maven.project.ProjectDependenciesResolver;
-import org.apache.maven.settings.Settings;
-import org.eclipse.aether.RepositorySystem;
-import org.eclipse.aether.RepositorySystemSession;
 import org.osgi.service.resolver.ResolutionException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import aQute.bnd.build.Workspace;
 import aQute.bnd.exporter.executable.ExecutableJarExporter;
 import aQute.bnd.exporter.runbundles.RunbundlesExporter;
-import aQute.bnd.maven.lib.configuration.BeanProperties;
-import aQute.bnd.maven.lib.configuration.Bndruns;
-import aQute.bnd.maven.lib.configuration.Bundles;
-import aQute.bnd.maven.lib.resolve.DependencyResolver;
+import aQute.bnd.maven.lib.resolve.ResolveMojo;
 import aQute.bnd.maven.lib.resolve.Scope;
+import aQute.bnd.maven.lib.resolve.TriConsumer;
 import aQute.bnd.osgi.Constants;
 import aQute.bnd.osgi.JarResource;
-import aQute.bnd.osgi.Processor;
 import aQute.bnd.osgi.Resource;
-import aQute.bnd.repository.fileset.FileSetRepository;
-import aQute.bnd.service.RepositoryPlugin;
 import aQute.lib.io.IO;
 import biz.aQute.resolve.Bndrun;
 import biz.aQute.resolve.ResolveProcess;
 
 @Mojo(name = "export", defaultPhase = PACKAGE, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
-public class ExportMojo extends AbstractMojo {
+@SuppressWarnings("deprecation")
+public class ExportMojo extends ResolveMojo {
```

Comment:
I'm somewhat concerned by this - "has a" relationships are almost always better than "is a" relationships. In this case an Export task isn't a resolve task, it just might need one.

Reply:
And again with the SuppressWarnings on the class declaration. I really dislike this since it will hide future warning we should take action on.

Reply:
There's no other way in Maven plugin model to support inheritance. What do you propose? To keep the over 80% duplication?

Reply:
Aggregation?

Reply:
For example, see how maven-jar-plugin uses maven-archiver. That is aggregation.

Reply:
hmm... I could look into this but it will not be possible to maintain compatibility as we'll have to move all shared parameters to a sub element of the configuration.

Reply:
I don't think you need to do that. Leave the configuration on the mojo and pass the fields to the aggregation which has the shared logic. This is what maven-jar-plugin does.

Reply:
You mean leave the 13 common parameters and 3 common components still spread across each Mojo? What does that buy us?

Reply:
The whole point started because all 3 mojos were constantly out of sync as far as their parameters. By using a super type we consolidate all of that into one place.

Reply:
It buys us aggregation instead of inheritance. Aggregation is always the first best choice since it provides maximum freedom to the change the implementation.

Reply:
So, I don't think that what you guys are suggesting is possible, not while also maintaining backward compatibility. Please point me at what you think is an implementation of this because so far I can't find one. The examples pointed out don't appear to be solutions.

Reply:
For example, Define a DTO class which has all the common fields. The execute can copy the mojo fields into a DTO object and pass the DTO object to the common resolve operation.

Or, for example, define a mojo base class which has getters for all the common fields and then pass the mojo (`this`) to the common resolve operation.

Reply:
@bjhargrave I'm well aware of how that could be designed. What I'm suggesting is that does NOT solve the repeated 13+3 fields that are at issue here.

Reply:
so you're suggesting throwing out the Parameter annotations...

Reply:
Agree, but I don't care about repeating the fields. I care about shared logic. The DTO/base getters class can enforce the commonality in the fields.

Reply:
Repeat the fields. Or maybe this is telling us, we need to common configuration type for resolving to transition to.

Reply:
The repeating fields is where we constantly mess up!

Reply:
Which is why maybe we need to introduce a common resolving configuration type and transition to use it in the various plugins which support resolving.